<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [68]:
import pandas as pd
import psycopg2

In [69]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
# удалено согласно условиям подачи проекта на проверку

In [70]:
connection = psycopg2.connect(
   dbname=DBNAME,
   user=USER,
   host=HOST,
   password=PASSWORD,
   port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [71]:
# вывожу подсчёт всех строк в таблице вакансий
query_3_1 = f'''SELECT
                COUNT(*) vacancies
                FROM vacancies
            '''

In [72]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


vacancies
0      49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [73]:
# вывожу подсчёт всех строк в таблице работодателей
query_3_2 = f'''SELECT
                COUNT(*) EMPLOYERS
                FROM EMPLOYERS
            '''

In [74]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\3588252056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


employers
0      23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [75]:
# вывожу подсчёт всех строк в таблице регионов
query_3_3 = f'''SELECT
                COUNT(*) AREAS
                FROM AREAS
            '''


In [76]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\2018224393.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


areas
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [77]:
# вывожу подсчёт всех строк в таблице сфер деятельности
query_3_4 = f'''SELECT
                COUNT(*) INDUSTRIES
                FROM INDUSTRIES
            '''


In [78]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\2172479966.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


industries
0         294

***

выводы по предварительному анализу данных:

Мы имеем дело с базой данных, в которой 23501 работодатель разместил 49197 вакансий

Покрытие базы - 1362 города

Взято во внимание 294 сферы деятельности

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [79]:
# вывожу регионы и подсчёт количества вакансий. Группирую по региону.
query_4_1 = f'''SELECT
                    a.name area,
                    COUNT(v.id) cnt
                FROM 
                    vacancies v
                JOIN areas a ON v.area_id = A.id
                GROUP BY area
                ORDER BY 2 DESC
            '''


In [80]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [81]:
# вывожу подсчёт вакансий и условие, чтобы либо верхняя либо нижняя граница зп была не пустая.
query_4_2 = f'''SELECT
                    COUNT(id) vacancies
                FROM 
                    vacancies 
                WHERE salary_from IS NOT NULL
                    OR salary_to IS NOT NULL
            '''

In [82]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\2168503709.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


vacancies
0      24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [83]:
# вывожу 2 столбца со средней верхней и средней нижней границей, округляю.
query_4_3 = f'''SELECT
                    ROUND(AVG(salary_from)) salary_from,
                    ROUND(AVG(salary_to)) salary_to
                FROM 
                    vacancies     
            '''

In [84]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\2222372203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


salary_from  salary_to
0      71065.0   110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [85]:
# вывожу столбцы: рабочий график, тип трудоустройства, подсчёт количества вакансий.
# группирую по 1 и 2 столбцу, чтобы выбрать только уникальные сочетания. Сортирую по убыванию количесва вакансий.
query_4_4 = f'''SELECT
                    schedule,
                    employment,
                    COUNT(id)
                FROM 
                    vacancies 
                GROUP BY 1,2
                ORDER BY 3 DESC               
            '''


In [86]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [87]:
# вывожу столбцы опыт и подсчёт вакансий. Группирую по опыту, сортирую по вакансиям.
query_4_5 = f'''SELECT
                    experience,
                    COUNT(id)
                FROM 
                    vacancies 
                GROUP BY 1
                ORDER BY 2                
            '''

In [88]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\2601166264.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# выводы по детальному анализу вакансий

Города, имеющие самое большое количество вакансий - Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. 
В конце своей работы я посмотрю на эти города в разрезе уровня предлагаемых заработных плат.

У 24073 вакансий заполнено хотя бы одно из двух полей с зарплатой. Эти данные очень важны, тк кандидаты обращают большее внимание на вакансии с уровнем оплаты.

Cредние значения для нижней и верхней границы зарплатной вилки: 71065 и 110537 рублей соответственно.

Самое популярное сочетание: тип рабочего графика - полный день и занятость - полная занятость. На втором месте по популярности - удалённая работа и полная занятость.
Вполне ожидаемо, что работодатели более заинтересованы в постоянном полноценном сотрудничестве. А удалённая работа, это тип рабочего графика, набирающий сейчас особую популярность.

Реже всего работодатели запрашивают опыт более 6 лет, далее порядке возрастания запросов: Нет опыта, от 3 до 6 лет и от 1 года до 3 лет.
На мой взгляд, работодатели редко ищут кандидатов с опытом более 6 лет по многим причинам. Им необходимо более сильная материальная мотивация, бОльшую часть работы можно выполнять
сотрудниками с меньшим опытом. Чаще всего это высокие должности, либо узконаправленные специальности.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [89]:
# вывожу столбцы с названиями компаний и подсчётом опубликованных вакансий. 
# Присоединяю таблицу вакансий. Группирую по компаниям.
# сортирую по количеству вакансий. Вывожу топ 5

query_5_1 = f'''SELECT
                    e.name employers,
                    COUNT(v.id) vacancies
                FROM 
                    employers e
                JOIN vacancies v ON e.id = v.employer_id
                GROUP BY 1
                ORDER BY 2 DESC
                LIMIT 5     
            '''

In [90]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\1792814825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


employers  vacancies
0         Яндекс       1933
1     Ростелеком        491
2       Тинькофф        444
3           СБЕР        428
4  Газпром нефть        331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [91]:
# вывожу столбцы - регион и подсчёт количества работодателей.
# объединяю таблицы регионов, работодателей и вакансий
# фильтрую по регионам, где нет вакансий. группирую по регионам, сортирую по кол-ву работодателей
query_5_2 = f'''SELECT 
                    areas.name area, 
                    COUNT(DISTINCT employers.id) employers_count
                FROM areas
                LEFT JOIN employers ON employers.area = areas.id
                LEFT JOIN vacancies ON vacancies.area_id = employers.area
                WHERE vacancies.id is null
                GROUP BY areas.name
                ORDER BY employers_count DESC        
            '''          

In [92]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\3341481329.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area  employers_count
0                   Россия              410
1                Казахстан              207
2       Московская область               75
3       Краснодарский край               19
4                 Беларусь               18
..                     ...              ...
889                Жабинка                0
890  Железногорск-Илимский                0
891                Жетысай                0
892               Жигалово                0
893               Жирновск                0

[894 rows x 2 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [93]:
# вывожу - компании и подсчёт регионов. Объединяю компании, вакансии, регионы.
# фильтрую, чтобы не было отсутствия данных. Группирую по компаниям, сортирую по количеству.

query_5_3 = f'''SELECT 
                    employers.name employers,
                    count(distinct(vacancies.area_id)) count_areas
                FROM employers
                JOIN vacancies ON employer_id = employers.id
                JOIN areas ON employers.area = areas.id
                WHERE vacancies.id is not null
                GROUP BY employers.name
                ORDER BY 2 desc
                            
            '''           

In [94]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


employers  count_areas
0                      Яндекс          181
1                  Ростелеком          152
2                  Спецремонт          116
3      Поляков Денис Иванович           88
4                   ООО ЕФИН            71
...                       ...          ...
14761                  UniSol            1
14762            UNISTORY LLC            1
14763                   UNIT6            1
14764     United Distribution            1
14765              UNITEL LLC            1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [95]:
# вывожу подсчёт работодателей. Объединяю работодателей с сферы деятельности.
# фильтрую по отсутствию сферы деятельности
query_5_4 = f'''SELECT
                    count(employers.id) employers
                from employers
                full join EMPLOYERS_INDUSTRIES on employers.id = employer_id
                where industry_id is null           
            '''           

In [96]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\3114411492.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


employers
0       8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [97]:
# вывожу название компаний, подсчёт их сфер деятельности
# объединяю компании и сферы деятельности. Группирую по сферам деятельности.
# фильтрую по количеству (4 сферы деятельности). Сортирую по алфавиту
query_5_5 = f'''SELECT
                    employers.name employers,
                    count(industry_id) count_industry
                FROM employers
                FULL JOIN EMPLOYERS_INDUSTRIES on employers.id = employer_id
                FULL JOIN INDUSTRIES on INDUSTRIES.id = INDUSTRY_id
                GROUP BY employers.name  
                HAVING count(industry_id) = 4     
                ORDER BY 1
                 
            '''           

In [98]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\1650419958.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


employers  count_industry
0        101 Интернет               4
1            21vek.by               4
2                2ГИС               4
3                  2К               4
4         4 пикселя +               4
...               ...             ...
1131            ЮРИОН               4
1132  ЮТИП Технологии               4
1133           ЯКласс               4
1134            ЯрНео               4
1135    Яшин&Партнёры               4

[1136 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [99]:
# вывожу только счётчик компаний. Объединяю работодателей и сферы деятельности
# фильтрую по контретной сфере деятельности.
query_5_6 = f'''SELECT
                    count(employers.name) count_employers                
                FROM employers
                FULL JOIN EMPLOYERS_INDUSTRIES on employers.id = employer_id
                FULL JOIN INDUSTRIES on INDUSTRIES.id = INDUSTRY_id 
                WHERE industries.name = 'Разработка программного обеспечения'                                   
            '''           

In [100]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\1169039133.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count_employers
0             3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [101]:
# импортирую необходимые библиотеки
import requests 
from bs4 import BeautifulSoup
import urllib.parse as req 

# присваиваю переменным ссылки и запросы. Я очень долго решала эту задачу и по способу, описанному в юните и у меня не получалось.
# поэтому я воспользовалась подсказкой, опубликованной кураторами. Следующие несколько строк из неё
path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров') 
url = path_common.scheme + '://' + req.quote(path_common.netloc) + req.quote(path_common.path) + '#'+ req.quote(path_common.fragment) 

response = requests.get(url) # Выполняю GET-запрос, присваиваю переменной ответ этого запроса

page = BeautifulSoup(response.text, 'html.parser')
# создаю кортеж с названиями городов из таблицы на сайте
mln_cities = tuple(a.text for a in page.find('table', class_="standard sortable").find_all('a')  if a.text[0].isupper())

# вывожу их, чтобы убедиться в правильности
print(mln_cities)



('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [102]:
# создаю таблицу, которая собирает вакансии в каждом городе миллионнике
# объединяю таблицы с регионами, вакансиями и работодателями, группирую по региону. Фильтрую по городам из кортежа.
# присоединяю к первой выборке, выборку из 1 строки. В ней я считаю все Вакансии Яндекса, так же применяя фильтр городов милионников
# сортирую по количеству вакансий

query_5_7 = f'''SELECT
                    areas.name as "Город миллионник", 
                    count(vacancies.id) as "Количество вакансий"
                FROM areas
                JOIN vacancies on areas.id = area_id
                JOIN employers on employers.id = employer_id
                WHERE employers.name = 'Яндекс'
                GROUP BY areas.name
                HAVING areas.name in {mln_cities}
                UNION 
                SELECT
                    'Total',
                    count(vacancies.id)
                FROM employers
                JOIN  vacancies on employers.id = employer_id
                JOIN areas on areas.id = area_id
                WHERE areas.name in {mln_cities}
                GROUP BY employers.name
                HAVING employers.name = 'Яндекс'
                ORDER BY 2
            '''           

In [103]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


Город миллионник  Количество вакансий
0              Омск                   21
1         Челябинск                   22
2        Красноярск                   23
3         Волгоград                   24
4             Пермь                   25
5            Казань                   25
6    Ростов-на-Дону                   25
7               Уфа                   26
8            Самара                   26
9         Краснодар                   30
10          Воронеж                   32
11      Новосибирск                   35
12  Нижний Новгород                   36
13     Екатеринбург                   39
14  Санкт-Петербург                   42
15           Москва                   54
16            Total                  485

# Выводы по анализу работодателей:

1) По количеству вакансий в пятёрку лидеров вошли работодатели: Яндекс, Ростелеком, Тинькофф, СБЕР и Газпром нефть. Это не удивительно.
Все перечисленные компании - крупнейшие в России, и , безусловно, им необходимо бОльшее число работников.
2) Регион, в котором наибольшее количество работодателей, но при этом нет вакансий - Россия. Думаю, что это связано с тем, что компании могут регистрировать
личный кабинет на сайте и при этом указывать регион своей компании - Россию, и это будет правильным. Особенно, если филиалы расположены в разных городах. 
Но когда работодатель размещает вакансию, необходимо указывать локацию для будущего сотрудника и в вакаснию вносят город.
3) Максимальное количество регионов, с опубликованными вакансиями 181 у компании Яндекс
4) 8419 работодателей не указали свою сферу деятельности. Что, однозначно, уменьшает их шансы быть замеченными кандидатами.
5) На 3м месте в списке компаний, указавших 4 сферы деятельности - 2ГИС. С учётом сортировки по алфавиту.
6) У 3553 работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».
7) Компания Яндекс разместила свои вакансии во всех 16 городах миллионниках. Всего это 485 вакансий

***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [104]:
# вывожу только подсчёт вакансий, фильтрую по всем возможным упоминаниям данных

query_6_1 = f'''SELECT
                    count(vacancies.name) vacancies
                FROM vacancies 
                WHERE vacancies.name LIKE '%данн%'
                    OR vacancies.name LIKE '%Данн%'
                    OR vacancies.name LIKE '%data%'
                    OR vacancies.name LIKE '%Data%'
                    OR vacancies.name LIKE '%DATA%'
                    OR vacancies.name LIKE '%ДАНН%'    
            '''           

In [105]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\485020324.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


vacancies
0       1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста ? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [106]:
# вывожу только подсчёт вакансий, фильтрую по всем возможным упоминаниям данных
# добавляю фильтры для молодого специалиста
query_6_2 = f'''SELECT
                    count(*) vacancies
                FROM vacancies 
                WHERE 
                    (name ILIKE '%data scientist%'
                    OR name ILIKE '%data science%'
                    OR name ILIKE '%исследователь данных%'
                    OR (name ILIKE '%ML%' and name not ilike '%HTML%')
                    OR name ILIKE '%machine learning%'
                    OR name ILIKE '%машинн%обучен%')  
                    and 
                    (name ILIKE '%junior%' 
                    OR experience ILIKE 'Нет опыта' 
                    OR employment ILIKE '%Стажировка%') 
            '''           

In [107]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\3641650667.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


vacancies
0         51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [108]:
# дублирую запрос из предыдущего задания, добавляю доп фильтр 
query_6_3 = f'''SELECT
                    count(*)
                from vacancies 
                where 
                    (name ILIKE '%data scientist%'
                    OR name ILIKE '%data science%'
                    OR name ILIKE '%исследователь данных%'
                    OR (name LIKE '%ML%' and name not ILIKE '%HTML%')
                    OR name ILIKE '%machine learning%'
                    OR name ILIKE '%машинн%обучен%')  
                    and
                    (key_skills ILIKE '%SQL%'
                    OR key_skills ILIKE '%postgres%')
            '''           


In [109]:

df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\3182346689.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [110]:
# дублирую запрос из предыдущего задания, добавляю доп фильтр
query_6_4 = f'''SELECT
                    count(*)
                from vacancies 
                where 
                    (name ILIKE '%data scientist%'
                    OR name ILIKE '%data science%'
                    OR name ILIKE '%исследователь данных%'
                    OR (name LIKE '%ML%' and name not ILIKE '%HTML%')
                    OR name ILIKE '%machine learning%'
                    OR name ILIKE '%машинн%обучен%')  
                    and
                    key_skills like '%Python%'
            '''

In [111]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [112]:
# из количества символов в строке с ключевыми навыками удаляю количество символов, которое образовалось при удалении разделителей
# прибавляю 1, тк разделителей меньше чем навыков. вычисляю среднее арифметическое, округляю до сотых
query_6_5 = f'''SELECT 
                    round(AVG(LENGTH(v.key_skills) - LENGTH(replace(v.key_skills,CHR(9),''))),2) + 1 skills
                from vacancies v
                where 
                    (name ILIKE '%data scientist%'
                    OR name ILIKE '%data science%'
                    OR name ILIKE '%исследователь данных%'
                    OR (name LIKE '%ML%' and name not ILIKE '%HTML%')
                    OR name ILIKE '%machine learning%'
                    OR name ILIKE '%машинн%обучен%')
                    AND 
                    v.key_skills IS NOT NULL  
                    
            '''

In [113]:

df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\980511775.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


skills
0    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [114]:
# блок с фильтрацией взяла из предыдущих заданий
# добавила фильтр, чтоб хотябы одно поле с зп было заполнено
# вывела столбцы с опытом и средней зп, использовала coalesce, чтобы, при отсутствии одной из ЗП, возвращал оставшуюся
# сгруппировала по опыту
query_6_6 = f'''SELECT 
                    experience, 
                    round(avg(coalesce((v.salary_from + v.salary_to)/2, v.salary_from, v.salary_to))) salary
                from vacancies v
                where 
                    (name ILIKE '%data scientist%'
                    OR name ILIKE '%data science%'
                    OR name ILIKE '%исследователь данных%'
                    OR (name LIKE '%ML%' and name not ILIKE '%HTML%')
                    OR name ILIKE '%machine learning%'
                    OR name ILIKE '%машинн%обучен%')
                    and (salary_to is not null or salary_from is not null)
                group by experience
                    '''

In [115]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience    salary
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

# выводы по предметному анализу
1) В нашем анализе к данным имеют отношение 1771 вакансия

2) Для начинающего дата-сайентиста найдена 51 вакансия

3) SQL или postgres в ключевых навыках указали работодатели в 201 вакансии

4) А Python, как ключевой навык хотели бы видеть в 351 вакансии

5) Более 6 ключевых навыков в среднем указывают в каждой вакансии DS

6) Для кандидатов без опыта вакансии предлагаются с зп 74643 руб в среднем, от 1 года жо 3х лет - 139675 руб, от 3х до 6 лет - 243115 руб.


Очевидно и очень показательно, что с увеличением опыта работы, работодатели готовы увеличивать и оплату труда. Задачи становятся труднее, но дороже.
То есть за 3 года ЗП специалиста может вырасти более, чем в 3 раза.

Но при этом, начинающему дата-сайентисту лучше не останавливаться на 1-2 навыках, необходимо расширять границы своего обучения, ведь работодатели хотят видеть специалистов широкого профиля.

# Общий вывод по проекту

Давайте посмотрим, в каких регионах дата-сайентисту работать "выгоднее". Ведь не все привязаны к месту и погут рассмотреть переезд в другой город, за более выгодными условиями.

Для этого, сделаю выборку из регионов, в которых опубликовано больше 15 вакансий (чтобы можно было выбрать из работодателей), посчитаю среднюю предложенную зп (с помощью своего более раннего кода) и отсортирую по уровню зп.

In [116]:
# вывожу 3 столбца с городом, количеством вакансий и средней зп
# фильтрую по "дата-вакансиям" и количеству вакансий
# сортирую
query_7_1 = f'''SELECT
                    a.name area,
                    count(v.id) vacancies,
                    round(avg(coalesce((v.salary_from + v.salary_to)/2, v.salary_from, v.salary_to))) salary
                FROM 
                    vacancies v
                JOIN areas a ON v.area_id = A.id
                 where v.name LIKE '%данн%'
                    OR v.name LIKE '%Данн%'
                    OR v.name LIKE '%data%'
                    OR v.name LIKE '%Data%'
                    OR v.name LIKE '%DATA%'
                    OR v.name LIKE '%ДАНН%'   
               
                GROUP BY area
                having round(avg(coalesce((v.salary_from + v.salary_to)/2, v.salary_from, v.salary_to))) > 0
                and count(v.id) > 15
                ORDER BY 3 DESC
            '''
df = pd.read_sql_query(query_7_1, connection)
df                   


C:\Users\Администратор\AppData\Local\Temp\ipykernel_1676\328915658.py:23: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


area  vacancies    salary
0           Армения         19  238906.0
1            Москва        619  173793.0
2       Новосибирск         57  154386.0
3      Екатеринбург         35  141000.0
4             Пермь         17  137500.0
5    Ростов-на-Дону         20  127500.0
6        Нур-Султан         55  122291.0
7   Санкт-Петербург        164  122112.0
8           Тбилиси         17  111249.0
9           Воронеж         20   91463.0
10           Самара         24   89290.0
11        Краснодар         36   86278.0
12           Казань         53   85000.0
13  Нижний Новгород         42   84538.0
14      Владивосток         16   84333.0
15          Ташкент         35   79522.0
16           Алматы        105   77976.0
17            Минск         89   75467.0
18            Томск         17   49758.0

Внезапно, для меня лично, Армения "готова предложить" соискателю самые высокие условия оплаты труда, это 238906 рублей в среднем по 19 вакансиям. Но при этом само количество вакансий в 30 раз ниже, чем в Москве.

Москва находится на втором месте по уровню зп (173к), и имеет при этом самое большое количество вакансий в нашей области (619).

На третьем месте Новосибирск. Уровень зп - выше 150000 рублей и количество вакансий достаточно высокое в нашей выборке - 57.

Интересно, что Санкт-Петербург попал только на 8 место, но при этом, может предложить соискателям 164 вакансии. Это 2 место по количеству в нашем анализе.

Так же, из того, на что необходимо обратить внимание - город Алматы. 3е место по количеству вакансий, но по уровню зп почти в самом конце. 

Выводы по анализу:


Мы имеем дело с базой данных, в которой 23501 работодатель разместил 49197 вакансий
Покрытие базы - 1362 города
Взято во внимание 294 сферы деятельности
Города, имеющие самое большое количество вакансий - Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. 

У 24073 вакансий заполнено хотя бы одно из двух полей с зарплатой. Эти данные очень важны, тк кандидаты обращают большее внимание на вакансии с уровнем оплаты.
Cредние значения для нижней и верхней границы зарплатной вилки: 71065 и 110537 рублей соответственно.
Самое популярное сочетание: тип рабочего графика - полный день и занятость - полная занятость. На втором месте по популярности - удалённая работа и полная занятость.
Вполне ожидаемо, что работодатели более заинтересованы в постоянном полноценном сотрудничестве. А удалённая работа, это тип рабочего графика, набирающий сейчас особую популярность.
Реже всего работодатели запрашивают опыт более 6 лет, далее порядке возрастания запросов: Нет опыта, от 3 до 6 лет и от 1 года до 3 лет.
На мой взгляд, работодатели редко ищут кандидатов с опытом более 6 лет по многим причинам. Им необходимо более сильная материальная мотивация, бОльшую часть работы можно выполнять
сотрудниками с меньшим опытом. Чаще всего это высокие должности, либо узконаправленные специальности.
По количеству вакансий в пятёрку лидеров вошли работодатели: Яндекс, Ростелеком, Тинькофф, СБЕР и Газпром нефть. Это не удивительно.
Все перечисленные компании - крупнейшие в России, и , безусловно, им необходимо бОльшее число работников.
Регион, в котором наибольшее количество работодателей, но при этом нет вакансий - Россия. Думаю, что это связано с тем, что компании могут регистрировать
личный кабинет на сайте и при этом указывать регион своей компании - Россию, и это будет правильным. Особенно, если филиалы расположены в разных городах. 
Но когда работодатель размещает вакансию, необходимо указывать локацию для будущего сотрудника и в вакаснию вносят город.
Максимальное количество регионов, с опубликованными вакансиями 181 у компании Яндекс

8419 работодателей не указали свою сферу деятельности. Что, однозначно, уменьшает их шансы быть замеченными кандидатами.
На 3м месте в списке компаний, указавших 4 сферы деятельности - 2ГИС. С учётом сортировки по алфавиту.
У 3553 работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».
Компания Яндекс разместила свои вакансии во всех 16 городах миллионниках. Всего это 485 вакансий
В нашем анализе к данным имеют отношение 1771 вакансия
Для начинающего дата-сайентиста найдена 51 вакансия
SQL или postgres в ключевых навыках указали работодатели в 201 вакансии
А Python, как ключевой навык хотели бы видеть в 351 вакансии
Более 6 ключевых навыков в среднем указывают в каждой вакансии DS
Для кандидатов без опыта вакансии предлагаются с зп 74643 руб в среднем, от 1 года жо 3х лет - 139675 руб, от 3х до 6 лет - 243115 руб.
Очевидно и очень показательно, что с увеличением опыта работы, работодатели готовы увеличивать и оплату труда. Задачи становятся труднее, но дороже.
То есть за 3 года ЗП специалиста может вырасти более, чем в 3 раза.
Но при этом, начинающему дата-сайентисту лучше не останавливаться на 1-2 навыках, необходимо расширять границы своего обучения, ведь работодатели хотят видеть специалистов широкого профиля.
